### DISCUSSION TITLE
Feed a pandas dataframe directly into Morph-KGC

### DISCUSSION THREAD

**Discussion by ToonTPubliq** [morph-kgc/morph-kgc#230] [Q&A] [answered]
Hi,
I want to update my code to take a pandas dataframe directly as input (together with a yarrrml-file) in morph-kgc. I am currently transforming my dataframe to a csv, and then give this csv to morph-kgc, but I think I can skip this step and just feed in the dataframe directly? Any help? 

This is the config and code I use now:

```python
config_ini = "[CONFIGURATION]
            output_format=N-QUADS

            [DataSource1]
            mappings= yarrrml-data/yarrrml.yml
            file_path: input-data/query_results.csv
            "

def generate_rdf_graph(config_ini_path):
    g = morph_kgc.materialize_oxigraph(config_ini_path)
    logging.debug("RDF graph generated successfully.")
    return g 
```

What exactly do I need to change to give a dataframe-variable instead of the csv-file as input?

Thanks for the help!!

**Comment by arenas-guerrero-julian** [comment, marked as answer]
Hi @ToonTPubliq ,

The documentation for in-memory data structures is [here](https://morph-kgc.readthedocs.io/en/latest/documentation/#rml-in-memory). You can also take a look at [these test cases](https://github.com/morph-kgc/morph-kgc/tree/main/test/rml-in-memory/pandas_dataframe).

CC @TheRazorace

**Reply by ToonTPubliq** [reply]
Hi @arenas-guerrero-julian ,
Thanks for your quick answer, but I am still struggling to implement the change. This also works with a yarrrml-file, right? I updated my mapping file and python code just like in the documentation, but I still get an error. Is this something in the morph-kgc library? Or am I doing something wrong.

I updated my "sources" parts in my yarrrml-file to: 
```python
sources:
      - [ "variable1", DataFrame]
```


And I updated my python code to ("kp_location_data" is my pandas dataframe):
```python
config_ini = f"""[CONFIGURATION]
            output_format=N-QUADS

            [DataSource1]
            mappings=location_yarrrml.yml
            """ 
data_dict = {"variable1": kp_location_data}
graph_store = generate_rdf_graph(config_ini, data_dict)
    
def generate_rdf_graph(config_ini, data_dict):
    g = morph_kgc.materialize_oxigraph(config_ini, data_dict)
    logging.debug("RDF graph generated successfully.")
    return g

```

But I still get this error:
```
    data = _preprocess_data(data, rml_rule, references, config)
                            ^^^^
UnboundLocalError: cannot access local variable 'data' where it is not associated with a value
```

Thanks for the help!!

**Reply by arenas-guerrero-julian** [reply]
RML in-memory is not currently supported for YARRRML, you have to use the tutle-based syntax.

**Reply by ToonTPubliq** [reply]
Ah okay, thanks! Is this something that might be added in the future?

**Reply by arenas-guerrero-julian** [reply]
It is not planned in the short term. If processing dataframes with Morph-KGC becomes popular I will give priority to it.
